File: 2020/aoc-elixir/lib/advent_of_code/day_06.ex
defmodule AdventOfCode.Day06 do
  def set_from_input(input) do
    String.split(input, ~r/\n\n/)
    |> Enum.map(fn x -> String.split(x)
                        |> Enum.map(&(String.to_charlist(&1)
                                      |> MapSet.new)) end)
  end
    
  def part1(args) do
    set_from_input(args)
    |> Enum.map(fn x -> Enum.reduce(x, &MapSet.union(&1, &2)) |> Enum.count end)
    |> Enum.sum
  end

  def part2(args) do
    set_from_input(args)
    |> Enum.map(fn x -> Enum.reduce(x, &MapSet.intersection(&1, &2)) |> Enum.count end)
    |> Enum.sum
  end
end

File: 2020/aoc-clojure-2020/src/aoc_clojure_2020/day_06.clj
(ns aoc-clojure-2020.day-06
  (:require [aoc-clojure-2020.helpers :refer [get-input]]
            [clojure.string :as s]
            [clojure.set :refer [union intersection]]))

(def input (as-> (get-input 6) i
             (s/split i #"\n\n")
             (map #(->> (s/split-lines %) (map set)) i)))

(def part-1 (reduce #(+ %1 (count (apply union %2))) 0 input))
(def part-2 (reduce #(+ %1 (count (apply intersection %2))) 0 input))
